### DISCUSSION TITLE
MakeDirectory 403 Access Denied.

### DISCUSSION THREAD

**Discussion by wadakatu** [laravel/framework#41656] [Q&A] [answered]
## Version

- PHP 8.1.3
- Laravel 9.4.1

## Problem

I have a problem with makeDirectory method.
Whenever I use makeDirectory method to create a directory on AWS S3 bucket, I get 403 error with the message of Access Denied.

```
Storage::disk('s3')->makeDirectory('main');  // 403 error (Access Denied)
```

However, if I use put method with file path and file name, it works without any error.
If there is no directory on S3 bucket, the method creates it automatically.

```
Storage::disk('s3')->put('main/test.txt', 'aaaa'); // works as I expected but why?
```

Is there anyone who had same experience that I have?
I do not think that I have authorization issue on AWS S3 bucket, 
because If I do, I get 403 error with put method as well.

Thank you.

**Comment by mikeydevelops** [comment, marked as answer]
Disclaimer: I don't use S3.

I presume you are using the [adapter suggested from the docs](https://laravel.com/docs/9.x/filesystem#s3-driver-configuration) for AWS and looking through the source code for Laravel and the adapter I see the difference is that the directory is made by [writing an empty string to the path you give it](https://github.com/thephpleague/flysystem-aws-s3-v3/blob/6f78e668ee70c69dd43830ebb21b41ec11f648fc/AwsS3V3Adapter.php#L272).
Maybe you do not have direct access to create directories?

**Reply by wadakatu** [reply]
@mikeydevelops 

Thank you for your comment.
I resolved the issue by myself.

The reason why 403 error is that there is no visibility option in my filesystem configuration.
After I added it, it works as I expected.

**Comment by etelford** [comment]
To clarify on the above comment "After I added it, it works as I expected"...

I had this same issue. I have a bucket that does not allow public objects. Before upgrading to Laravel 9, this worked:

```php
// config/filesystems.php
'my_s3_bucket' => [
    'driver' => 's3',
    'region' => env('AWS_REGION'),
    'bucket' => env('AWS_BUCKET'),
    'url' => env('AWS_URL'),
    'endpoint' => env('AWS_ENDPOINT'),
],
```

but after upgrading to Laravel 9 (with `"league/flysystem-aws-s3-v3": "^3.0",`) I had to add the following to `filesystems.php`:

```php
'options' => ['ACL' => 'private'],
```

so the final config looks like:

```php
'my_s3_bucket' => [
    'driver' => 's3',
    'region' => env('AWS_REGION'),
    'bucket' => env('AWS_BUCKET'),
    'url' => env('AWS_URL'),
    'endpoint' => env('AWS_ENDPOINT'),
    'options' => ['ACL' => 'private'],
],
```

**Reply by challgren** [reply]
Wouldn't just adding `'visibility' => 'private',` work? According to https://github.com/laravel/framework/blob/9.x/src/Illuminate/Filesystem/FilesystemManager.php#L237-L239 when the S3 driver is created the visbility is converted into an ACL by the League\Flysystem\AwsS3V3\AwsS3V3Adapter and League\Flysystem\AwsS3V3\PortableVisibilityConverter

**Reply by denys-rybas** [reply]
'options' => ['ACL' => 'private'] works for me, thanks!

**Reply by Sammyjo20** [reply]
I can confirm that setting `visibility` to `private` in the disks' config file worked for me 👍
